tools/plugin-commands-self-updater/src/selfUpdate.ts

```diff
@@ -0,0 +1,87 @@
+import fs from 'fs'
+import path from 'path'
+import { docsUrl } from '@pnpm/cli-utils'
+import { getCurrentPackageName, packageManager, isExecutedByCorepack } from '@pnpm/cli-meta'
+import { createResolver } from '@pnpm/client'
+import { pickRegistryForPackage } from '@pnpm/pick-registry-for-package'
+import { types as allTypes } from '@pnpm/config'
+import { PnpmError } from '@pnpm/error'
+import { getToolDirPath } from '@pnpm/tools.path'
+import { linkBins } from '@pnpm/link-bins'
+import pick from 'ramda/src/pick'
+import renderHelp from 'render-help'
+import { add, type InstallCommandOptions } from '@pnpm/plugin-commands-installation'
+
+export function rcOptionsTypes (): Record<string, unknown> {
+  return pick([], allTypes)
+}
+
+export function cliOptionsTypes (): Record<string, unknown> {
+  return {
+    ...rcOptionsTypes(),
+  }
+}
+
+export const commandNames = ['self-update']
+
+export function help (): string {
+  return renderHelp({
+    description: 'Updates pnpm to the latest version',
+    descriptionLists: [],
+    url: docsUrl('self-update'),
+    usages: [],
+  })
+}
+
+export type SelfUpdateCommandOptions = InstallCommandOptions
+
+export async function handler (
+  opts: SelfUpdateCommandOptions
+): Promise<undefined | string> {
+  if (isExecutedByCorepack()) {
+    throw new PnpmError('CANT_SELF_UPDATE_IN_COREPACK', 'You should update pnpm with corepack')
+  }
+  const { resolve } = createResolver({ ...opts, authConfig: opts.rawConfig })
+  const pkgName = 'pnpm'
+  const resolution = await resolve({ alias: pkgName, pref: 'latest' }, {
+    lockfileDir: opts.lockfileDir ?? opts.dir,
+    preferredVersions: {},
+    projectDir: opts.dir,
+    registry: pickRegistryForPackage(opts.registries, pkgName, 'latest'),
+  })
+  if (!resolution?.manifest) {
+    throw new PnpmError('CANNOT_RESOLVE_PNPM', 'Cannot find latest version of pnpm')
+  }
+  if (resolution.manifest.version === packageManager.version) {
+    return 'Already the latest version is installed'
+  }
+
+  const currentPkgName = getCurrentPackageName()
+  const dir = getToolDirPath({
+    pnpmHomeDir: opts.pnpmHomeDir,
+    tool: {
+      name: currentPkgName,
+      version: resolution.manifest.version,
+    },
+  })
+  if (fs.existsSync(dir)) {
+    await linkBins(path.join(opts.dir, opts.modulesDir ?? 'node_modules'), opts.pnpmHomeDir,
+      {
+        warn: () => {},
+      }
+    )
+    return `Latest version is already present on the system. Linked from ${opts.dir}`
+  }
+  fs.mkdirSync(dir, { recursive: true })
+  fs.writeFileSync(path.join(dir, 'package.json'), '{}')
+  await add.handler(
+    {
+      ...opts,
+      dir,
+      lockfileDir: dir,
+      bin: opts.pnpmHomeDir,
+    },
+    [`${currentPkgName}@${resolution.manifest.version}`]
+  )
+  return undefined
```

Comment:
Do we eventually need to clean up older versions of pnpm installed in this manner? It might be good to do this here, and delete installs older than 30 days?

Reply:
We don't know if there are projects on the computer still using the older versions. pnpm allows to define a specific pnpm version via the packageManager field of `package.json`.